### ISSUE TITLE
Variants / sum types / tagged unions / discriminated unions / etc.

### ISSUE THREAD

**Issue by sinelaw** [cling-lang/spec#1]
Maybe github issues are better than wiki pages for discussions?
Anyway, we could make a rule that extensions can always with very small ifdefs get back to standard C *that behaves the same at runtime*.

So for variants,
```c
struct A { }
struct B { }

enum Tag { A, B };

struct VariantAB {
    union { struct A a; struct B b; } u;
    enum Tag tag;
};

// elsewhere:
struct VariantAB x;
switch (x.tag) {
case A: // ... use x.u.a
    break;
case B: // ... use x.u.b
    break;
}
```

So far valid standard C. Whatever changes we make should be minor alterations on top of that.

**Comment by sinelaw** [comment]
Here's an idea:
```C

struct VariantAB {
    union {
        tag(A) struct A a; // note the 'tag(A)'
        tag(B) struct B b;
    }
    enum Tag tag; // standard c
};
```
and then code of the form: `switch (x.tag) { case ... ` will be verified such that in each case, you are only allowed to access the union member that has the corresponding `tag(X)` attribute.


```

**Comment by da-x** [comment]
Yeah, we can use issues, and update the Wiki page when we reach conclusions.

The problem I see with `switch (x.tag) { case .. ` is that it is ambiguous whether you really intent to switch on the enum value or over the union types, unless we modify the `case ...` part somehow to resolve the ambiguity. Maybe `case (A : a)`, meaning that `a` is our inner `struct A` and if there was `switch (x->tag)` then a will be of type `struct A *`?

As for the struct definition in your suggestion (missing `u ;`?), it cannot be `#ifdef`'ed out easily. I'd like the ability to turn structs in Cling headers back to C structs with a simple `#ifdef` like the example in the Wiki page. I think we should also make sure not to restrict the memory layout of the struct, i.e. preserve C's memory layout flexibility by default.

**Comment by da-x** [comment]
BTw, I think the main concern with this feature is not adding full featured sum types, but to prevent the following bug (which currently only heavily configured static code analysis helps with):

E.g.:

```c
switch (x.tag) {
...
case A:
     printf("%d"\n", x.b.var); // OOPS!
     break;
...
}
```

This takes care of that simply by providing 'a' and using it instead of decoding the union by hand every time, like it is usually done by C programmers. There is another side of this bug in initializers too.

**Comment by sinelaw** [comment]
Not sure what you mean by the ambiguity. Switching over the enum tag *is* switching over the union types. 
Also, what's the problem with #ifdef-ing away? I gave an example of what you can do to make it a teenee tiny change (#definition of `tag(x)`) to make it fully compatibly with standard C.
I agree, the feature should only affect type checking.

**Comment by da-x** [comment]
Okay I referring to the introduction of 'a' to the scope, I gather that your proposition is to restrict access to the other union members. I think it is too hard to restrict access it's like private in C++, and it's too unlike C). But having 'a' is good enough, and constitutes good sugaring I think.

#Ifdefing away the `tag(A)` is more difficult than just surrounding `discrim {}` block with one `#ifdef`.

**Comment by sinelaw** [comment]
Why is it hard to restrict? You just fail compilation if any statement under the case accesses any but the valid union member. In fact you can disallow accessing ANY union members outside of switch cases, to enforce that you must pattern match to get a value. 
Of course one can still with pointers/casts get to internal members of the struct unsafely, but that's not something we should be trying to prohibit.

#ifdefing in a single header file the meaning of `tag` is easier than sprinkling #ifdefs all over your `discrim` blocks everywhere.

**Comment by sinelaw** [comment]
BTW, with the plain struct syntax + disallow wrong union, you get code that is EXACTLY the same as what you would do in standard C. The extension only adds extra level of static checking that would otherwise not be there.

**Comment by da-x** [comment]
It is hard to restrict in a sense that it require iterating over the whole expression tree of the `case`s in order to find sub expressions that might access the restricted union fields.

Here's an example for why we should not restrict access:

```c
struct {
    union {
         struct a;
         struct b;
         char raw[0x10]; // Here raw does not even have a tag, it's just for debugging */
    } u;

    enum t tag;
};

// Later in code
func() 
{
    ...
    switch (x.tag) {
    ...
    case B:
           /* Handling B, and debugging of the memory representation in that case, 
            * to chase down some segfault we are having */
           hexdump(x.u.raw, sizeof(x.u.raw));
           break;
           ....
    }
    ..
}
```

Restricting here would disallow some debugging machinery, otherwise useful, and I bet quite prevalent in existing C code.

Next, there's another reason why to feel unease regarding the matching of enum values to union fields inside the union like in your suggestion, rather than outside. Here's an example:

```c
union u_many {
    struct opt_a a;
    struct opt_b b;
    struct opt_c c;
};

struct x {
    union u_many u;
    enum tag_x {X_A, X_B} tag; /* limited to A and B */
};

struct y {
    union u_many u;
    enum tag_y {Y_B, Y_C} tag; /* limited to B and C */
};
```

This example should evince two things: 1) It's possible that the union is declared outside the struct, and in various existing code, it is done as such, not strictly bound to where you have the enum declared 2) the idea that the matching is a property of the tag rather than of the union itself.

Actually, it does not have to be union in term of memory representation if the user desires. The switch over the tag is only a map between a value (the tag) and sub expressions of various types receive a pointer to a struct (that happens to contain the tag) as a parameter. The tagged union idea is just a special case for that.

**Comment by da-x** [comment]
BTW I suppose we can have restricted union field access to be made optional with by enabling an additional extension?

**Comment by sinelaw** [comment]
I agree with everything you say, as we discussed. But, I think a simplistic extension that just checks your plain tagged unions in the way I described is useful for many (most?) use cases, and has the advantage of being just an attribute that can be ignored by a standard C compiler to produce exactly the code you would want anyway (but without the static safety checks). 

So yeah, maybe we want two extensions? One that can deal with arbitrary representation as you described, and another simpler one for basic tagged unions.

**Comment by da-x** [comment]
Okay, then to reiterate, we should have a `TagMappedSubFields` extension that extends the syntax in 'struct' and 'case' (introducing the matched field in a scope, i.e. access via `a`), and a 'TaggedUnions` extension that relies on potentially vanishing union field attributes, but does not modify any other aspect of the syntax (i.e. accesses are still via `x.u.a.`).

**Comment by sinelaw** [comment]
Sounds good

**Comment by da-x** [comment]
Okay, then let's have the extension's spec in Markdown files in the repository instead of the wiki (which seems redundant). Plus, we'll open issues and discuss about more extensions once they come up.

**Comment by sinelaw** [comment]
Hey, take a look at: http://clang.llvm.org/docs/AttributeReference.html#argument-with-type-tag
and the other attributes following in the docs

**Comment by da-x** [comment]
It doesn't mention any kind of struct field attribute

**Comment by da-x** [comment]
BTW, does the `<identifier>(A)` in your spec needs to match the field name 'enum Tag <identifier>', or is it a keyword, or do you assume that the name is 'tag'?

**Comment by sinelaw** [comment]
`tag(A)` is sugar for `__attribute__((variant_tag(A)))`
The actual tag field name has no effect.

BTW, turns out the syntax for attributes on fields/vars is after the var name.

```C
struct VariantAB {
    union {
         struct Foo a __attribute__((variant_tag(A))) ;
         struct Bar b __attribute__((variant_tag(B)));
    }
    enum MyTag my_tag; // standard c
};
```

**Comment by da-x** [comment]
Okay, so upon encountering such attribute, it would look up the enum type of the provided value, and then look up a field in the containing struct that matches that type, finding `my_tag` field in this case?

**Comment by da-x** [comment]
Updated https://github.com/cling-lang/spec/blob/master/TagMappedDesignation.md
